File: go/shipreal_test.go
package shipreal

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

// Fixtures deliberately mirror the sandbox: same shapes, frozen contents.
var moduleOne = Module{
	Slug: "sandbox-module-1", Title: "Sandbox: Thinking in Systems & State",
	Part: "Sandbox Part 1: The Mental Shift", Description: "A fixed fixture module.",
	Chapters: 12, Minutes: 29, URL: "https://example.invalid/sandbox-module-1",
}

func server(t *testing.T, handler http.HandlerFunc) (*Client, *httptest.Server) {
	t.Helper()
	srv := httptest.NewServer(handler)
	t.Cleanup(srv.Close)
	return New(WithBaseURL(srv.URL)), srv
}

func TestSearchSendsQueryAndParsesPage(t *testing.T) {
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		if got := r.URL.Path; got != "/api/v1/modules" {
			t.Errorf("path = %q", got)
		}
		if got := r.URL.Query().Get("q"); got != "caching" {
			t.Errorf("q = %q", got)
		}
		if got := r.URL.Query().Get("limit"); got != "50" {
			t.Errorf("limit = %q", got)
		}
		if !strings.HasPrefix(r.Header.Get("user-agent"), "shipreal-go/") {
			t.Errorf("user-agent = %q", r.Header.Get("user-agent"))
		}
		_ = json.NewEncoder(w).Encode(Page{Data: []Module{moduleOne}})
	})
	page, err := c.Search(context.Background(), SearchParams{Query: "caching", Limit: 50})
	if err != nil {
		t.Fatal(err)
	}
	if len(page.Data) != 1 || page.Data[0].Slug != moduleOne.Slug {
		t.Fatalf("data = %+v", page.Data)
	}
}

func TestSandboxRoutesUnderSandboxPrefix(t *testing.T) {
	var seen string
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		seen = r.URL.Path
		_ = json.NewEncoder(w).Encode(Page{})
	})
	c.sandbox = true
	if _, err := c.Search(context.Background(), SearchParams{}); err != nil {
		t.Fatal(err)
	}
	if seen != "/api/v1/sandbox/modules" {
		t.Fatalf("path = %q", seen)
	}
}

func TestModulesFollowsCursorAndStops(t *testing.T) {
	calls := 0
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		calls++
		cursor := r.URL.Query().Get("cursor")
		next := "page2"
		if cursor == "" {
			_ = json.NewEncoder(w).Encode(Page{
				Data:       []Module{moduleOne},
				Pagination: Pagination{HasMore: true, NextCursor: &next},
			})
			return
		}
		if cursor != "page2" {
			t.Errorf("cursor = %q", cursor)
		}
		_ = json.NewEncoder(w).Encode(Page{Data: []Module{{Slug: "second"}}})
	})
	all, err := c.Modules(context.Background(), "")
	if err != nil {
		t.Fatal(err)
	}
	if calls != 2 {
		t.Fatalf("calls = %d, want 2", calls)
	}
	if len(all) != 2 || all[1].Slug != "second" {
		t.Fatalf("modules = %+v", all)
	}
}

// An empty-string cursor has to terminate the loop as surely as a null one, or
// a server that sends one spins the client forever.
func TestModulesStopsOnEmptyCursor(t *testing.T) {
	calls := 0
	empty := ""
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		calls++
		_ = json.NewEncoder(w).Encode(Page{
			Data:       []Module{moduleOne},
			Pagination: Pagination{HasMore: true, NextCursor: &empty},
		})
	})
	if _, err := c.Modules(context.Background(), ""); err != nil {
		t.Fatal(err)
	}
	if calls != 1 {
		t.Fatalf("calls = %d, want 1", calls)
	}
}

func TestModuleEscapesTitleFragments(t *testing.T) {
	var seen string
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		seen = r.URL.EscapedPath()
		_ = json.NewEncoder(w).Encode(moduleOne)
	})
	if _, err := c.Module(context.Background(), "systems & state"); err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(seen, "%20&%20") && !strings.Contains(seen, "systems") {
		t.Fatalf("escaped path = %q", seen)
	}
}

func TestProblemDetailsBecomeError(t *testing.T) {
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("content-type", "application/problem+json")
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"type":"https://shipreal.dev/developers#errors-module-not-found",` +
			`"title":"Module Not Found","status":404,"detail":"No module matches \"nope\"."}`))
	})
	_, err := c.Module(context.Background(), "nope")
	var apiErr *Error
	if !errors.As(err, &apiErr) {
		t.Fatalf("err = %v (%T)", err, err)
	}
	if apiErr.Status != 404 {
		t.Errorf("status = %d", apiErr.Status)
	}
	if !strings.HasSuffix(apiErr.Type, "#errors-module-not-found") {
		t.Errorf("type = %q", apiErr.Type)
	}
	if !strings.Contains(apiErr.Error(), "Module Not Found") {
		t.Errorf("message = %q", apiErr.Error())
	}
}

// A non-JSON error body must not swallow the status, which is the only thing
// the caller can still act on.
func TestNonJSONErrorBodyKeepsStatus(t *testing.T) {
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadGateway)
		_, _ = w.Write([]byte("upstream fell over"))
	})
	_, err := c.Course(context.Background())
	var apiErr *Error
	if !errors.As(err, &apiErr) || apiErr.Status != 502 {
		t.Fatalf("err = %v", err)
	}
}

func TestPricingRegionPicksDeliberately(t *testing.T) {
	p := Pricing{
		Complete: Plan{Intl: RegionPrice{Now: "$99"}, IL: RegionPrice{Now: "₪299"}},
		Teams:    Plan{Intl: RegionPrice{Now: "$119"}, IL: RegionPrice{Now: "₪359"}, MinSeats: 1},
	}
	complete, seat, ok := p.Region("il")
	if !ok || complete.Now != "₪299" || seat.Now != "₪359" {
		t.Fatalf("il = %v %v %v", complete, seat, ok)
	}
	// An unknown region falls back to international and says so, rather than
	// quietly quoting a price for the wrong country.
	if _, _, ok := p.Region("uk"); ok {
		t.Fatal("unknown region reported ok")
	}
}

func TestBatchRefusesOverTheCeiling(t *testing.T) {
	c := New()
	reqs := make([]BatchRequest, MaxBatch+1)
	if _, err := c.Batch(context.Background(), reqs); err == nil {
		t.Fatal("want an error for an oversized batch")
	}
}

func TestBatchPostsRequests(t *testing.T) {
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			t.Errorf("method = %s", r.Method)
		}
		var body struct {
			Requests []BatchRequest `json:"requests"`
		}
		_ = json.NewDecoder(r.Body).Decode(&body)
		if len(body.Requests) != 1 || body.Requests[0].Path != "/pricing" {
			t.Errorf("requests = %+v", body.Requests)
		}
		_ = json.NewEncoder(w).Encode(BatchResponse{Count: 1, Responses: []BatchItem{{ID: "0", Status: 200}}})
	})
	out, err := c.Batch(context.Background(), []BatchRequest{{Path: "/pricing"}})
	if err != nil {
		t.Fatal(err)
	}
	if out.Count != 1 {
		t.Fatalf("count = %d", out.Count)
	}
}

func TestAskStreamYieldsEventsInOrder(t *testing.T) {
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("accept"); got != "text/event-stream" {
			t.Errorf("accept = %q", got)
		}
		w.Header().Set("content-type", "text/event-stream")
		_, _ = w.Write([]byte(
			"event: start\ndata: {\"_meta\":{\"response_type\":\"result_list\",\"count\":1}}\n\n" +
				"event: result\ndata: {\"results\":[{\"name\":\"Caching\"}]}\n\n" +
				"event: complete\ndata: {\"_meta\":{\"response_type\":\"complete\"}}\n\n"))
	})
	var events []string
	var names []string
	err := c.AskStream(context.Background(), "caching", func(e AskEvent) error {
		events = append(events, e.Event)
		for _, r := range e.Results {
			names = append(names, r.Name)
		}
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
	if strings.Join(events, ",") != "start,result,complete" {
		t.Fatalf("events = %v", events)
	}
	if len(names) != 1 || names[0] != "Caching" {
		t.Fatalf("names = %v", names)
	}
}

func TestAskStreamStopsWhenCallbackErrors(t *testing.T) {
	c, _ := server(t, func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte("event: start\ndata: {}\n\nevent: result\ndata: {}\n\n"))
	})
	stop := errors.New("enough")
	seen := 0
	err := c.AskStream(context.Background(), "caching", func(AskEvent) error {
		seen++
		return stop
	})
	if !errors.Is(err, stop) {
		t.Fatalf("err = %v", err)
	}
	if seen != 1 {
		t.Fatalf("callback ran %d times, want 1", seen)
	}
}

func TestEmptyArgumentsAreRejectedBeforeAnyRequest(t *testing.T) {
	c, srv := server(t, func(w http.ResponseWriter, r *http.Request) {
		t.Error("no request should have been sent")
	})
	_ = srv
	if _, err := c.Module(context.Background(), ""); err == nil {
		t.Error("Module accepted an empty argument")
	}
	if _, err := c.Ask(context.Background(), ""); err == nil {
		t.Error("Ask accepted an empty question")
	}
	if err := c.AskStream(context.Background(), "", func(AskEvent) error { return nil }); err == nil {
		t.Error("AskStream accepted an empty question")
	}
}

File: ruby/test/shipreal_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "json"
require_relative "server_helper"
require_relative "../lib/shipreal"

# A real socket rather than a stub, so the client's own HTTP handling (headers,
# query building, SSE framing, error parsing) is what gets exercised.
class ShipRealTest < Minitest::Test
  include TestServer

  MODULE_ONE = {
    "slug" => "sandbox-module-1", "title" => "Sandbox: Thinking in Systems & State",
    "part" => "Sandbox Part 1", "description" => "A fixed fixture module.",
    "chapters" => 12, "minutes" => 29, "url" => "https://example.invalid/sandbox-module-1"
  }.freeze

  JSON_HEADERS = { "content-type" => "application/json" }.freeze

  def json_ok(body)
    [200, JSON_HEADERS, JSON.generate(body)]
  end

  # Runs the block with a client pointed at a server driven by `handler`.
  def serving(handler)
    with_server(handler) do |base|
      yield ShipReal::Client.new(base_url: base)
    end
  end

  def test_search_sends_query_and_parses_page
    seen = nil
    serving(lambda { |req|
      seen = req
      json_ok({ "data" => [MODULE_ONE] })
    }) do |client|
      page = client.search(query: "caching", limit: 50)
      assert_equal 1, page["data"].length
      assert_equal MODULE_ONE["slug"], page["data"][0]["slug"]
    end
    assert_equal "/api/v1/modules", seen.path
    assert_equal "caching", seen.query["q"]
    assert_equal "50", seen.query["limit"]
    assert_match(/\Ashipreal-ruby\//, seen.headers["user-agent"])
  end

  def test_sandbox_routes_under_sandbox_prefix
    seen = nil
    with_server(lambda { |req|
      seen = req
      json_ok({ "data" => [] })
    }) do |base|
      ShipReal::Client.new(base_url: base, sandbox: true).search
    end
    assert_equal "/api/v1/sandbox/modules", seen.path
  end

  def test_modules_follows_cursor_then_stops
    calls = 0
    serving(lambda { |req|
      calls += 1
      if req.query["cursor"].nil?
        json_ok({ "data" => [MODULE_ONE],
                  "pagination" => { "hasMore" => true, "nextCursor" => "page2" } })
      else
        json_ok({ "data" => [{ "slug" => "second", "cursorSeen" => req.query["cursor"] }],
                  "pagination" => { "hasMore" => false, "nextCursor" => nil } })
      end
    }) do |client|
      all = client.modules
      assert_equal 2, all.length
      assert_equal "second", all[1]["slug"]
      assert_equal "page2", all[1]["cursorSeen"]
    end
    assert_equal 2, calls
  end

  # An empty-string cursor has to terminate the loop as surely as a nil one, or
  # a server that sends one spins the client forever.
  def test_modules_stops_on_empty_cursor
    calls = 0
    serving(lambda { |_req|
      calls += 1
      json_ok({ "data" => [MODULE_ONE],
                "pagination" => { "hasMore" => true, "nextCursor" => "" } })
    }) { |client| client.modules }
    assert_equal 1, calls
  end

  def test_module_escapes_the_path_segment
    seen = nil
    serving(lambda { |req|
      seen = req
      json_ok(MODULE_ONE)
    }) { |client| client.module_by("systems & state") }
    # Both the space and the ampersand are encoded. An unescaped & in a path
    # segment is the kind of thing that works until a title contains one.
    assert_equal "/api/v1/modules/systems%20%26%20state", seen.path
  end

  def test_problem_details_become_an_error
    serving(lambda { |_req|
      [404, { "content-type" => "application/problem+json" }, JSON.generate({
        "type" => "https://shipreal.dev/developers#errors-module-not-found",
        "title" => "Module Not Found", "status" => 404,
        "detail" => 'No module matches "nope".'
      })]
    }) do |client|
      err = assert_raises(ShipReal::Error) { client.module_by("nope") }
      assert_equal 404, err.status
      assert_match(/#errors-module-not-found\z/, err.type)
      assert_match(/No module matches/, err.message)
    end
  end

  # A non-JSON error body must not swallow the status, which is the only thing
  # the caller can still act on.
  def test_non_json_error_body_keeps_status
    serving(->(_req) { [502, {}, "upstream fell over"] }) do |client|
      err = assert_raises(ShipReal::Error) { client.course }
      assert_equal 502, err.status
      assert_nil err.type
    end
  end

  def test_pricing_region_flattens_deliberately
    body = {
      "free" => { "price" => 0, "includes" => "overviews" },
      "complete" => { "intl" => { "now" => "$99" }, "il" => { "now" => "₪299" }, "url" => "u" },
      "teams" => { "intl" => { "now" => "$792" }, "il" => { "now" => "₪2,392" }, "packSeats" => 10 }
    }
    serving(->(_req) { json_ok(body) }) do |client|
      il = client.pricing(region: "il")
      assert_equal "il", il["region"]
      assert_equal "₪299", il["complete"]["now"]
      assert_equal 10, il["teams"]["packSeats"]
      # No region named returns both, unflattened, rather than guessing one.
      both = client.pricing
      assert both["complete"].key?("intl")
      assert both["complete"].key?("il")
    end
  end

  def test_batch_refuses_over_the_ceiling
    client = ShipReal::Client.new
    assert_raises(ArgumentError) do
      client.batch(Array.new(ShipReal::MAX_BATCH + 1) { { "path" => "/pricing" } })
    end
  end

  def test_batch_posts_requests
    seen = nil
    serving(lambda { |req|
      seen = req
      json_ok({ "count" => 1, "responses" => [{ "id" => "0", "status" => 200 }] })
    }) do |client|
      out = client.batch([{ "path" => "/pricing" }])
      assert_equal 1, out["count"]
    end
    assert_equal "POST", seen.method
    assert_equal "/api/v1/batch", seen.path
    assert_equal [{ "path" => "/pricing" }], JSON.parse(seen.body)["requests"]
  end

  def test_ask_stream_yields_events_in_order
    seen = nil
    serving(lambda { |req|
      seen = req
      [200, { "content-type" => "text/event-stream" },
       "event: start\ndata: {\"_meta\":{\"count\":1}}\n\n" \
       "event: result\ndata: {\"results\":[{\"name\":\"Caching\"}]}\n\n" \
       "event: complete\ndata: {\"_meta\":{}}\n\n"]
    }) do |client|
      events = []
      names = []
      client.ask_stream("caching") do |ev|
        events << ev["event"]
        Array(ev.dig("data", "results")).each { |r| names << r["name"] }
      end
      assert_equal %w[start result complete], events
      assert_equal ["Caching"], names
    end
    assert_equal "text/event-stream", seen.headers["accept"]
  end

  def test_empty_arguments_are_rejected_before_any_request
    # Points at a closed port: reaching the network at all would be the failure.
    client = ShipReal::Client.new(base_url: "http://127.0.0.1:1")
    assert_raises(ArgumentError) { client.module_by("") }
    assert_raises(ArgumentError) { client.ask("") }
    assert_raises(ArgumentError) { client.ask_stream("") { nil } }
  end
end
